### PR TITLE
Tags CRUD endpoints

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3047238&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php [hunk 1]
```diff
@@ -55,95 +53,88 @@ public function __construct
...
-
-            $validation = Validator::make($values, $rules);
-
-            if ($validation->fails()) {
-                $ex = new ValidationException();
-                throw $ex->setMessages($validation->messages()->toArray());
+        return $this->_getAll(
+            function () {
+                return [
+                    'tag' => ['=@', '=='],
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php:59]: lets add '@@' as additional operator

File: app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php [hunk 2]
```diff
@@ -55,95 +53,88 @@ public function __construct
...
-            if (Request::has('page')) {
-                $page     = intval(Request::input('page'));
-                $per_page = intval(Request::input('per_page'));
-            }
+            if(is_null($tag)) return $this->error404();
 
-            $filter = null;
+            return $this->ok(SerializerRegistry::getInstance()
+                ->getSerializer($tag)
+                ->serialize(Request::input('expand', '')));
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php:97]: use serializeUtils to provide all serialize method params

File: app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php [hunk 3]
```diff
@@ -158,26 +149,21 @@ public function addTag(){
                 );
             }
 
-            $tag = $this->tag_service->addTag($data->all());
+            $tag = $this->tag_service->updateTag(intval($tag_id), $data->all());
 
-            return $this->created(SerializerRegistry::getInstance()->getSerializer($tag)->serialize
+            return $this->updated(SerializerRegistry::getInstance()->getSerializer($tag)->serialize
             (
                 Request::input('expand','')
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php:156]: use serializeUtils same as above

File: app/Services/Model/Imp/TagService.php [hunk 4]
```diff
@@ -58,4 +59,35 @@ public function addTag(array $payload): Tag
...
+     */
+    public function updateTag(int $tag_id, array $payload): Tag
+    {
+        return $this->tx_service->transaction(function () use ($tag_id, $payload){
+            $tag = $this->tag_repository->getById($tag_id);
+            if(is_null($tag) || !$tag instanceof Tag){
+                throw new EntityNotFoundException("tag {$tag_id} not found");
+            }
+
+            $tag->setTag(trim($payload['tag']));
```
Comment (smarcet) [inline app/Services/Model/Imp/TagService.php:74]: u need check here for dupes before add and throw a validation error

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php (hunk 1 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php:59]: done

File: app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php (hunk 3 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/OAuth2TagsApiController.php:156]: done

File: app/Services/Model/Imp/TagService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/TagService.php]: done

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**